libraries/AP_Proximity/AP_Proximity_Backend.cpp

```diff
@@ -79,6 +82,83 @@ bool AP_Proximity_Backend::ignore_reading(uint16_t angle_deg) const
             }
         }
     }
+
+    // check if obstacle is near land
+    if (check_obstacle_near_ground(angle_deg, distance_m)) {
+        return true;
+    }
+
+    return false;
+}
+
+// store rangefinder values
+void AP_Proximity_Backend::set_rangefinder_alt(bool use, bool healthy, float alt_cm)
+{
+    _last_downward_update_ms = AP_HAL::millis();
+    _rangefinder_use = use;
+    _rangefinder_healthy = healthy;
+    _rangefinder_alt = alt_cm * 0.01f;
+}
+
+// get alt from rangefinder in meters
+bool AP_Proximity_Backend::get_rangefinder_alt(float &alt_m) const
+{
+    if (!_rangefinder_use || !_rangefinder_healthy) {
+        // range finder is not healthy
+        return false;
+    }
+
+    const uint32_t dt = AP_HAL::millis() - _last_downward_update_ms;
+    if (dt > PROXIMITY_ALT_DETECT_TIMEOUT_MS) {
+        return false;
+    }
+
+    // readings are healthy
+    alt_m = _rangefinder_alt;
+    return true;
+}
+
+// Check if Obstacle defined by body-frame yaw and pitch is near ground
+bool AP_Proximity_Backend::check_obstacle_near_ground(float yaw, float pitch, float distance) const
+{
+    if (!frontend._ign_gnd_enable) {
+        return false;
+    }
+    if (!hal.util->get_soft_armed()) {
+        // don't run this feature while vehicle is disarmed, otherwise proximity data will not show up on GCS
+        return false;
+    }
+    // Assume object is yaw and pitch bearing and distance meters away from the vehicle
+    Vector3f object_3D;
+    object_3D.offset_bearing(wrap_180(yaw), wrap_180(-pitch), distance);
```

Comment:
I don't think we can use wrap_180 on the pitch.. so I think we should remove it or throw it away if the pitch is >90 or less than -90.

Reply:
Done. Fixed this in master at another place also